docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+
+Admission controllers will watch for this operation in order to approve the
+images from the modified pod spec. Containers can be added regardless of pod
+status to allow for troubleshooting containers in states like
+`CrashLoopBackoff`.
+
+##### Alternative: Create a New Container Type
+
+Rather than modifying the default list of containers, we could create an
+additional list of containers in the spec similar to `InitContainers`. This
+would reduce the complexity of modifying the default containers list, but we
+should avoid a proliferation of container types unless there's a compelling
+reason. A mutable container list is a better long-term goal even if it is more
+difficult.
+
+#### Container Volumes
+
+Covered by [#831](https://issues.k8s.io/831), we'll want to be able to make
+volumes and eventually root filesystems of the other containers in the pod
+available to the troubleshooting container. This is a complex issue and covered
+in a separate proposal.
+
+When `kubectl` generates a patch to add a container to the pod spec, it will
+include configuration to do the equivalent of a "volumes from" the other
+containers in the pod, depending on how [#831](https://issues.k8s.io/831) is
+implemented.
+
+#### Container runOnce
+
+In order to support short-lived processes (e.g. troubleshooting sessions,
+automation) from the general pod spec container list we need to be able to
+override the pod's `restartPolicy` for the newly added container so it is not
+restarted when it exits. When a "run once" container exits, it will transition
+to the Terminated state, not be restarted, and continue to assert the `Ready`
+condition.
+
+![Pod State Diagram](images/pod-troubleshooting-state.png "Pod State Diagram")
+
+The container will continue to be part of the pod spec, allowing automation to
+inspect the pod, but its resources could be freed by the container runtime. The
+container could be started again manually by performing an update on the pod
+that re-adds the container.
+
+### `kubectl exec` a Container
+
+Once these other features have been implemented, we will add an additional
+option to `kubectl exec` which will cause it to patch a running pod, wait for
+the new container to start and attach to its console. This means it will be able
+to reattach to the troubleshooting session using a standard `kubectl attach`.
+
+```
+% kubectl exec -it -m gcr.io/proj/debug nginx bash
+Reattach to this session using: kubectl attach -it -c debug nginx
+If you don't see a command prompt, try pressing enter.
+root@nginx:/#
+```
+
+#### Alternative: Create New `debug` Command
+
+Rather than extending `kubectl exec` we could create `kubectl debug` to handle
+this use case (and others?). This would be similar to things like `oc debug` and
+could fit into longer term troubleshooting strategy, but it's not technically
+necessary. A new command might be less confusing for someone expecting `kubectl
+exec` to mirror `docker exec` exactly.
```

Comment:
This alternative requires 0% new conceptual changes to pods which I think is worth pointing out in the proposal.  "Adding a container to pod spec" and "Container runOnce" are huge changes to the system and thus have to meet a proportionally higher bar of change.

Can you articulate other use cases for "container runOnce" and "adding a container to pod spec" beyond debugging that justify the expense of that change?  How could "container runOnce" be abused?  I consider that within the scope of this proposal to answer and defend against those points.